source/common/common/radix_tree.h

```diff
@@ -0,0 +1,324 @@
+   * @return pointer to the child node, or nullptr if not found.
+   */
+  const RadixTreeNode* getChild(const RadixTreeNode& node, uint8_t char_key) const {
+    auto it = node.children_.find(char_key);
+    if (it != node.children_.end()) {
+      return &it->second;
+    }
+    return nullptr;
+  }
+
+  /**
+   * Check if a node has a value (is a leaf node)
+   */
+  static bool hasValue(const RadixTreeNode& node) {
+    // For pointer types, check if the pointer is not null
+    if constexpr (std::is_pointer_v<Value>) {
+      return node.value_ != nullptr;
+    } else {
+      return static_cast<bool>(node.value_);
+    }
+  }
+
+  /**
+   * Find the longest common prefix between two strings
+   */
+  static size_t longestPrefix(absl::string_view a, absl::string_view b) {
+    size_t len = std::min(a.size(), b.size());
+    for (size_t i = 0; i < len; i++) {
+      if (a[i] != b[i]) {
+        return i;
+      }
+    }
+    return len;
+  }
+
+public:
+  /**
+   * Adds an entry to the RadixTree at the given Key.
+   * @param key the key used to add the entry.
+   * @param value the value to be associated with the key.
+   * @param overwrite_existing will overwrite the value when the value for a given key already
+   * exists.
+   * @return false when a value already exists for the given key.
+   */
+  bool add(absl::string_view key, Value value, bool overwrite_existing = true) {
+    // Check if the key already exists
+    Value existing = find(key);
+
+    // If a value exists and we shouldn't overwrite, return false
+    if (static_cast<bool>(existing) && !overwrite_existing) {
+      return false;
+    }
+
+    root_.insert(key, key, std::move(value));
+    return true;
+  }
+
+  /**
+   * Finds the entry associated with the key.
+   * @param key the key used to find.
+   * @param result the value associated with the key (only set if found).
+   * @return true if the key was found, false otherwise.
+   */
+  bool find(absl::string_view key, Value& result) const {
+    return root_.find_recursive(key, result);
+  }
+
+  /**
+   * Finds the entry associated with the key.
+   * @param key the key used to find.
+   * @return the Value associated with the key, or an empty-initialized Value
+   *         if there is no matching key.
+   */
+  Value find(absl::string_view key) const {
+    Value result;
+    if (find(key, result)) {
+      return result;
+    }
+    return Value{};
+  }
+
+  /**
+   * Returns the set of entries that are prefixes of the specified key, longest last.
+   * Complexity is O(min(longest key prefix, key length)).
+   * @param key the key used to find.
+   * @return a vector of values whose keys are a prefix of the specified key, longest last.
+   */
+  absl::InlinedVector<Value, 4> findMatchingPrefixes(absl::string_view key) const {
+    absl::InlinedVector<Value, 4> result;
+    absl::string_view search = key;
+    const RadixTreeNode* node = &root_;
+    bool consumed_prefix = false;
```

Comment:
I think `consumed_prefix` is unnecessary and actually makes for incorrect behavior - if the tree was constructed like
```
tree.add("", "x");
tree.add("/", "y");
tree.add("/foo", "z");
```
findMatchingPrefixes would incorrectly return ["y", "z"] instead of ["x", "y", "z"] because of this check.

Also after removing that check, the special case below can be removed too, as the loop will then provide identical behavior.

This comment also applies to findLongestPrefix.